### PR TITLE
[SPARK-3165][MLlib] DecisionTree use sparsity in data

### PR DESCRIPTION
**Pull request by facaiy**
## What changes were proposed in this pull request?
For store efficiency, 
use sparse representation when number of feature >= 10000;
use dense otherwise.

## How was this patch tested?

+ [x] pass all previous unit tests.
+ [x] add unit test to trigger sparse condition.
+ [ ] Some performance benchmark perhaps are need.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala [hunk 1]
```diff
@@ -301,7 +302,7 @@ private[tree] class LearningNode(
    *         group of nodes on one call to
    *         [[org.apache.spark.ml.tree.impl.RandomForest.findBestSplits()]].
    */
-  def predictImpl(binnedFeatures: Array[Int], splits: Array[Array[Split]]): Int = {
+  def predictImpl(binnedFeatures: Int => Int, splits: Array[Array[Split]]): Int = {
```
Comment (facaiy) [inline mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala]: when use `breeze.linalg.SparseVector`, compile fails:
```scala
java.lang.AssertionError: assertion failed: List(method apply$mcI$sp, method apply$mcI$sp)
        at scala.reflect.internal.Symbols$Symbol.suchThat(Symbols.scala:1947)
        at scala.tools.nsc.transform.SpecializeTypes$$anon$2.matchingSymbolInPrefix$1(SpecializeTypes.scala:1474)
        at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transformSelect$1(SpecializeTypes.scala:1497)
        at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1593)
        at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$anonfun$transform$3.apply(SpecializeTypes.scala:1442)
        ......
```

I don't know why? Can anyone help? 
Thanks.

Comment (facaiy) [inline mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala]: when use `breeze.linalg.SparseVector`, compile fails:
```scala
java.lang.AssertionError: assertion failed: List(method apply$mcI$sp, method apply$mcI$sp)
        at scala.reflect.internal.Symbols$Symbol.suchThat(Symbols.scala:1947)
        at scala.tools.nsc.transform.SpecializeTypes$$anon$2.matchingSymbolInPrefix$1(SpecializeTypes.scala:1474)
        at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transformSelect$1(SpecializeTypes.scala:1497)
        at scala.tools.nsc.transform.SpecializeTypes$$anon$2.transform1(SpecializeTypes.scala:1593)
        at scala.tools.nsc.transform.SpecializeTypes$$anon$2$$anonfun$transform$3.apply(SpecializeTypes.scala:1442)
        ......
```

I don't know why? Can anyone help? 
Thanks.

**Review by WeichenXu123 — COMMENTED** [review]
@facaiy I want to ask you two questions before starting review this:
1) Do you make some perf test for this and how much improvement on very large dimensions sparse vector ?
Currently the tree algo do not have good performance on large dimensions training. But the sparsity case is usually large dimensions.
2) Do you find some scenario which must use decision tree training on large dimension sparse vector ?

Thanks! also cc @jkbradley

**Comment by facaiy** [comment]
Thank you for comment. 

Very good question, at least for me, the answer to both questions is no. In most case, we feed dense raw data into tree model. However, if large dimensions required, I believe the sparse representation is necessary.

**Comment by WeichenXu123** [comment]
@facaiy So can you do benchmark first (by generating random testing data) ?  So we can see how much this can speed up.

**Comment by facaiy** [comment]
Sure, @WeichenXu123 , perhaps one or two weeks later, is it OK?

By the way, I think using sparse representation can only reduce memory usage, and it is in the cost of compute performance. Hence, it cannot speed up calculation for low-dimension data. The only advantage sparse representation has is for super large dimension in my opinion.

**Comment by facaiy** [comment]
Hi, since the work has been done for a long time, I take a review by myself. 

After careful review, as SparseVector is compressed sparse row format, so the only benefit of the PR would be for data storage but in the cost of performance. But for tree-method, it is uncommon to handle a super large dimension features. Hence, it cannot satisfy me.

I prefer to [SPARK-3717: DecisionTree, RandomForest: Partition by feature](https://issues.apache.org/jira/browse/SPARK-3717) as an alternative, which will be benefits in both performance and storage if I understand correctly. So the PR is closed. Thank everyone for review / comment.

**Pull request closed without merging**